### ISSUE TITLE
Jams object completeness

### ISSUE THREAD

**Issue by rabitt** [mir-dataset-loaders/mirdata#226]
For ease of contribution, we've implemented a `jams_utils.jams_converter` function which writes jams objects under the hood for each Track object's `.to_jams()` method. This has two benefits:
* contributors do not need to be familiar with the jams api
* each dataset module's code is drastically simplified.

However, this limits the completeness of the resulting jams object - namely, we omit much of the metadata that could be included. We now enforce that the jams objects returned by `Track.to_jams()` passes validation, so they are minimally usable, but they are not ideal.

Do we have any ideas for how to improve the jams object completeness without hurting the ease of contribution? Do we need fully complete jams files for most use cases (e.g. usage with `muda`)

**Comment by lostanlen** [comment]
I don't take sides on the strict completeness of JAMS, but i'll say this: i wish our `to_jams` required a `duration` input as first argument instead of computing it from the Track `self`.

Since #212, this works:
```
from mirdata import gtzan_genre, ikala

dataset = gtzan_genre
with tempfile.TemporaryDirectory() as dirpath:
    track_id = dataset.track_ids()[0]
    track = dataset.Track(track_id, data_home=dirpath)
    jam = track.to_jams()
    assert jam.validate()
    print("Yay, Track is JAMS-compliant!")
```

This doesn’t:
```
dataset = ikala
with tempfile.TemporaryDirectory() as dirpath:
    track_id = dataset.track_ids()[0]
    track = dataset.Track(track_id)
    jam = track.to_jams()
    assert jam.validate()
    print("Yay, Track is JAMS-compliant!")
```

Here's my backtrace:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/librosa/librosa/core/audio.py in load(path, sr, mono, offset, duration, dtype, res_type)
    126     try:
--> 127         with sf.SoundFile(path) as sf_desc:
    128             sr_native = sf_desc.samplerate

~/miniconda3/lib/python3.7/site-packages/soundfile.py in __init__(self, file, mode, samplerate, channels, subtype, endian, format, closefd)
    739                                          format, subtype, endian)
--> 740         self._file = self._open(file, mode_int, closefd)
    741         if set(mode).issuperset('r+') and self.seekable():

~/miniconda3/lib/python3.7/site-packages/soundfile.py in _open(self, file, mode_int, closefd)
   1264         _error_check(_snd.sf_error(file_ptr),
-> 1265                      "Error opening {0!r}: ".format(self.name))
   1266         if mode_int == _snd.SFM_WRITE:

~/miniconda3/lib/python3.7/site-packages/soundfile.py in _error_check(err, prefix)
   1454         err_str = _snd.sf_error_number(err)
-> 1455         raise RuntimeError(prefix + _ffi.string(err_str).decode('utf-8', 'replace'))
   1456 

RuntimeError: Error opening '/var/folders/dt/q6wyf56s41lbv8z8cr0147t00000gp/T/tmp711ps2oe/Wavfile/10161_chorus.wav': System error.

During handling of the above exception, another exception occurred:

FileNotFoundError                         Traceback (most recent call last)
<ipython-input-19-e524265e0709> in <module>
      3     track_id = dataset.track_ids()[0]
      4     track = dataset.Track(track_id, data_home=dirpath)
----> 5     jam = track.to_jams()
      6     assert jam.validate()

~/mirdata/mirdata/ikala.py in to_jams(self)
    141                 'track_id': self.track_id,
    142                 'song_id': self.song_id,
--> 143                 'duration': librosa.get_duration(self.mix_audio[0], self.mix_audio[1]),
    144             },
    145         )

~/mirdata/mirdata/ikala.py in mix_audio(self)
    129     def mix_audio(self):
    130         """(np.ndarray, float): mono mixture audio signal, sample rate"""
--> 131         return load_mix_audio(self.audio_path)
    132 
    133     def to_jams(self):

~/mirdata/mirdata/ikala.py in load_mix_audio(audio_path)
    189 
    190     """
--> 191     mixed_audio, sr = librosa.load(audio_path, sr=None, mono=True)
    192     # multipy by 2 because librosa averages the left and right channel.
    193     return 2.0 * mixed_audio, sr

~/librosa/librosa/core/audio.py in load(path, sr, mono, offset, duration, dtype, res_type)
    142         if isinstance(path, (str, pathlib.PurePath)):
    143             warnings.warn('PySoundFile failed. Trying audioread instead.')
--> 144             y, sr_native = __audioread_load(path, offset, duration, dtype)
    145         else:
    146             raise(exc)

~/librosa/librosa/core/audio.py in __audioread_load(path, offset, duration, dtype)
    166 
    167     y = []
--> 168     with audioread.audio_open(path) as input_file:
    169         sr_native = input_file.samplerate
    170         n_channels = input_file.channels

~/miniconda3/lib/python3.7/site-packages/audioread/__init__.py in audio_open(path)
     78     from . import rawread
     79     try:
---> 80         return rawread.RawAudioFile(path)
     81     except DecodeError:
     82         pass

~/miniconda3/lib/python3.7/site-packages/audioread/rawread.py in __init__(self, filename)
     59     """
     60     def __init__(self, filename):
---> 61         self._fh = open(filename, 'rb')
     62 
     63         try:

FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/dt/q6wyf56s41lbv8z8cr0147t00000gp/T/tmp711ps2oe/Wavfile/10161_chorus.wav'
```

**Comment by rabitt** [comment]
Good catch on the `ikala` error - in fact I think any module where the audio is not on disk will error. (I wish jams objects did not require duration in the first place...)

Having the `to_jams` methods take `duration` as an argument is a nice solution, but it's of course slightly harder to use (and error prone), since the user will have to get the per-track duration themselves. How about the following compromise - in the modules where `to_jams()` infers duration from Track.audio (when Track.duration does not exist), we:
* pass `duration=None` by default: `to_jams(duration=None)`
* when `duration` is specified we use it directly
* we create a `utils.get_duration(audio_path)` method, which is called when `duration=None`. When the audio path does not exist, we raise a helpful error that will break before the jams object is created.

This way, the old API usage doesn't change and people can simply call `.to_jams()` and it will work. If it doesn't work, we're providing a more helpful error message than before, and the user has a option to manually specify it if they want. What do you think @lostanlen ?

**Comment by lostanlen** [comment]
> Having the to_jams methods take duration as an argument is a nice solution, but it's of course slightly harder to use (and error prone), since the user will have to get the per-track duration themselves.

@rabitt exactly. 

> This way, the old API usage doesn't change and people can simply call .to_jams() and it will work.

Honestly, i don't think people who use mirdata v0.1 mind replacing their `.to_jams()` function by `to_jams(duration)`. Having `duration=None` as a default value might cause silent bugs in user code. I would prefer to throw a cold `TypeError` to the face of our (yet small) JAMS crowd, which they can fix in a few minutes by writing a custom duration estimator, than to risk silent bugs. I don't think mirdata is in a mature state yet to estimate the duration of a Track efficiently and reliably.

**Comment by rabitt** [comment]
Yes but this substantially reduces usability - as users will have to do exactly what we are already doing - estimate the duration from the track's audio file.

I don't see any risk of silent bugs - see this comment:
> we create a utils.get_duration(audio_path) method, which is called when duration=None. When the audio path does not exist, we raise a helpful error that will break before the jams object is created.

So this will replace the existing non-silent  bug with a more useful, also non-silent bug.

**Comment by lostanlen** [comment]
@rabitt 
```
from mirdata import ikala
track_id = ikala.track_ids()[0]
jam = ikala.to_jams()
if jam.duration is None:
    jam.duration = my_duration_estimator()
```

If you run this code in mirdata v0.1: the code passes and executes my_duration_estimator()
If you run this code on current master, the code passes and executes some mirdata internal function

The two values may not be the same. Therefore, it's a silent bug.

**Comment by rabitt** [comment]
I agree the previous behavior (before #212 ) was a bug, namely because the `jam` object did not validate. 

It seems to me there are two issues:
1) if the audio file is not available, `to_jams()` throws a hard to parse error
2) we want the resulting jam.file_metadata.duration value to be consistent

On (1), this is where I think the ideal behavior is to catch this case and return a more helpful error. This can be done easily by creating, for example, a `utils.get_duration(filepath)` method that will handle this case.

On (2), if we allow `duration` to be a parameter, doesn't this exactly break (2)? Then we're leaving it up to the user to get the "correct" duration value. I would argue that we want to define one and exactly one "correct" duration. Otherwise, if `duration` is a mandatory argument
```python
from mirdata import ikala
track_id = ikala.track_ids()[0]
track = ikala.Track(track_id)
jam1 = track.to_jams(my_correct_custom_duration)
jam2 = track.to_jams(my_incorrect_custom_duration)
```
Would results in two jams objects corresponding to the same audio file but with different durations. For me this is way more silent error prone than defining duration within `to_jams()`.

Going back to your orignal stack trace, as I understand the reasons for adding an explict duration argument is so that `to_jams` could be used whether or not the audio is available. If we want to support this, this is where I would suggest `duration=None` as the default, which would allow users to get a jams object if the audio is not available, but this is still error prone, as they will still need to somehow get the duration themselves, and this could be done "wrong".

So, my vote is to address (1) as I suggested above, and to _not_ require duration in the `to_jams()` call.

**Comment by lostanlen** [comment]
@rabitt well, regarding the code you gave, the user can always do
```
assert jam1.file_metadata.duration == jam2.file_metadata.duration
```
if they suspect a mismatch. But the mismatch will be totally their fault anyway, because they computed the values `my_correct_custom_duration` and `my_incorrect_custom_duration` with a custom program.

The problem i'm raising is not that two JAMS objects describing the same Track but different durations would coexist in user code. JAMS objects are mutable, just like any Python object. So there's no way to prevent people to shoot themselves in the foot if they write code which estimates duration in several different ways. That's a case where we can blame the user.

My concern is that updating the mirdata package from beta to master, _without changing user code_, will lead to a different value of `jam.file_metadata.duration` for the user.

**Comment by magdalenafuentes** [comment]
I don't like the idea of asking the user to compute the track duration and pass it as a parameter on jams, it goes against simplifying usability of `to_jams()`, and also I'm not following the explanation of the two possible durations, so I don't get why would it be a silent bug, @lostanlen can you elaborate a bit please? Otherwise it feels too much discussion for a minor issue.

I like the solution of throwing a hard to parse error if audio files are not available and computing the duration ourselves if (if `duration` not available as metadata).

**Comment by rabitt** [comment]
> My concern is that updating the mirdata package from beta to master, without changing user code, will lead to a different value of jam.file_metadata.duration for the user.

Ok, this is different than I understood you were saying. I don't personally have an issue with updating a buggy, non-validating jams file with `None` as the duration (in beta) to a validating hans fuke where duration is specified (in master).

If the value of duration were specified (not None) in both cases, and _changing_ then I of course have an issue, but it's not the case.
